easybuild/tools/modules.py

```diff
@@ -131,6 +133,116 @@
 _log = fancylogger.getLogger('modules', fname=False)
 
 
+class ModuleEnvironmentVariable:
+    """Environment variable data structure for modules"""
+    def __init__(self, paths, top_level_file=False):
+        self.paths = paths
+        self.top_level_file = bool(top_level_file)
+
+    def __str__(self):
+        return ":".join(self.paths)
+
+    @property
+    def paths(self):
+        return self._paths
+
+    @paths.setter
+    def paths(self, value):
+        """Enforce that paths is a list of strings"""
+        if isinstance(value, str):
+            value = [value]
+        try:
+            self._paths = [str(path) for path in value]
+        except TypeError:
+            raise TypeError("ModuleEnvironmentVariable.paths must be a list of strings") from None
+
+    def append(self, *args):
+        """Shortcut to append to list of paths"""
+        self.paths.append(*args)
+
+    def extend(self, *args):
+        """Shortcut to extend list of paths"""
+        self.paths.extend(*args)
+
+    def prepend(self, item):
+        """Shortcut to append to list of paths"""
+        self.paths.insert(0, item)
+
+
+# singleton metaclass: only one instance is created
+BaseModuleEnvironment = create_base_metaclass('BaseModuleEnvironment', Singleton, object)
```

Comment:
@lexming Using a singleton here (so a single instance that is shared across all `EasyBlock` instances) seems like a bad idea to me...

If EasyBuild installs two things in the same session (via `--robot`, or by being fed multiple easyconfigs on the command line), you'll end up with incorrect stuff in the generated module file due to this, no?

Reply:
@lexming I've implemented a test that shows that using a singleton here is a bad idea, see https://github.com/lexming/easybuild-framework/pull/5